### PR TITLE
Proxy error reworking

### PR DESCRIPTION
**Pull request by conradludgate**
## Problem

Taking my ideas from https://github.com/neondatabase/neon/pull/6283 and doing a bit less radical changes. smaller commits.

We currently don't report error classifications in proxy as the current error handling made it hard to do so.

## Summary of changes

1. Add a `ReportableError` trait that all errors will implement. This provides the error classification functionality.
2. Handle Client requests a strongly typed error
    * this error is a `ReportableError` and is logged appropriately
3. The handle client error only has a few possible error types, to account for the fact that at this point errors should be returned to the user.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: proxy/src/serverless/conn_pool.rs [hunk 1]
```diff
@@ -261,6 +265,23 @@ pub struct GlobalConnPoolOptions {
     pub max_total_conns: usize,
 }
 
+#[derive(Debug, thiserror::Error)]
```
Comment (khanova) [inline proxy/src/serverless/conn_pool.rs:268]: I would suggest to move it to `../backend.rs` and rename it to `HttpConnError` or something.

The idea is to separate the connection logic from pool.

File: proxy/src/cancellation.rs [hunk 2]
```diff
@@ -1,24 +1,42 @@
...
+    #[error("{0}")]
+    IO(#[from] std::io::Error),
+    #[error("{0}")]
+    Postgres(#[from] tokio_postgres::Error),
+}
+
+impl ReportableError for CancelError {
+    fn get_error_type(&self) -> crate::error::ErrorKind {
+        match self {
+            CancelError::IO(_) => crate::error::ErrorKind::Compute,
```
Comment (khanova) [inline proxy/src/cancellation.rs:26]: Isn't it `Disconnected`?

File: proxy/src/error.rs [hunk 3]
```diff
@@ -29,7 +29,7 @@ pub trait UserFacingError: fmt::Display {
     }
 }
 
-#[derive(Clone)]
+#[derive(Copy, Clone, Debug)]
 pub enum ErrorKind {
```
Comment (khanova) [inline proxy/src/error.rs:33]: On the second thoughts, I think, it's rather `reason`, than `kind`, but I am not sure. WDYT?

File: proxy/src/serverless/sql_over_http.rs [hunk 4]
```diff
@@ -170,36 +173,27 @@ fn get_conn_info(
...
     .await;
     let mut response = match result {
         Ok(r) => match r {
-            Ok(r) => r,
+            Ok(r) => {
+                ctx.set_success();
+                r
+            }
             Err(e) => {
+                // ctx.set_error_kind(e.get_error_type());
```
Comment (khanova) [inline proxy/src/serverless/sql_over_http.rs:195]: Remove?

File: proxy/src/serverless/sql_over_http.rs [hunk 5]
```diff
@@ -287,6 +283,9 @@ pub async fn handle(
             )?
         }
     };
+
+    ctx.log();
```
Comment (khanova) [inline proxy/src/serverless/sql_over_http.rs:287]: Hm, actually, why did you change it to explicit events reporting (instead of doing it within the destructor)?

**Review by khanova — APPROVED** [review]
Looks good! Left a few minor questions.

File: proxy/src/error.rs (hunk 3 above)
Comment (conradludgate) [reply proxy/src/error.rs:33]: I think both options seem fine

File: proxy/src/cancellation.rs (hunk 2 above)
Comment (conradludgate) [reply proxy/src/cancellation.rs:26]: I will rename `Disconnect` to `ClientDisconnect` to make the distinction clearer. Currently `Compute` covers both errors returned by postgres + disconnections from compute. Would you like me to also separate them into `Compute` and `ComputeDisconnect`?

File: proxy/src/serverless/sql_over_http.rs (hunk 5 above)
Comment (conradludgate) [reply proxy/src/serverless/sql_over_http.rs:287]: Yeah it's not necessary. left over from a prior change. Will remove

File: proxy/src/serverless/sql_over_http.rs (hunk 4 above)
Comment (conradludgate) [reply proxy/src/serverless/sql_over_http.rs]: marked as todo
